File: backend/app/Services/CartService.php
<?php

namespace App\Services;

use App\Models\Cart;
use App\Models\CartItem;
use App\Models\Product;
use App\Models\User;
use Illuminate\Validation\ValidationException;

class CartService
{
    public function getOrCreateCart(User $user): Cart
    {
        return Cart::firstOrCreate(['user_id' => $user->id]);
    }

    /**
     * Tambah item ke keranjang dengan pengecekan stok SERVER-SIDE
     * (docs/16: "Stock must be checked server-side.").
     */
    public function addItem(User $user, array $data): CartItem
    {
        $product = Product::active()->findOrFail($data['product_id']);

        $requestedQty = $data['quantity'] ?? 1;
        $available = $this->availableStock($product);

        $existing = $this->getOrCreateCart($user)->items()
            ->where('product_id', $product->id)
            ->where('variant_id', $data['variant_id'] ?? null)
            ->first();

        $newQty = $requestedQty + ($existing->quantity ?? 0);
        if ($newQty > $available) {
            throw ValidationException::withMessages([
                'quantity' => ["Stok tidak mencukupi. Tersisa {$available}."],
            ]);
        }

        // Harga diambil dari database — jangan pernah percaya harga dari klien.
        $unitPrice = $this->unitPrice($product, $data['variant_id'] ?? null);

        if ($existing) {
            $existing->update(['quantity' => $newQty, 'unit_price' => $unitPrice]);

            return $existing;
        }

        return $this->getOrCreateCart($user)->items()->create([
            'product_id' => $product->id,
            'variant_id' => $data['variant_id'] ?? null,
            'quantity' => $newQty,
            'unit_price' => $unitPrice,
        ]);
    }

    public function updateItem(User $user, CartItem $cartItem, int $quantity): CartItem
    {
        $this->authorizeCartItem($user, $cartItem);

        if ($quantity < 1) {
            throw ValidationException::withMessages(['quantity' => ['Jumlah minimal 1.']]);
        }

        $available = $this->availableStock($cartItem->product);
        if ($quantity > $available) {
            throw ValidationException::withMessages(['quantity' => ["Stok tidak mencukupi. Tersisa {$available}."]]);
        }

        $cartItem->update(['quantity' => $quantity]);

        return $cartItem;
    }

    public function removeItem(User $user, CartItem $cartItem): void
    {
        $this->authorizeCartItem($user, $cartItem);
        $cartItem->delete();
    }

    public function clear(User $user): void
    {
        $this->getOrCreateCart($user)->items()->delete();
    }

    public function show(User $user): Cart
    {
        return $this->getOrCreateCart($user)->load('items.product:id,name,slug,price,stock', 'items.variant');
    }

    public function subtotal(Cart $cart): float
    {
        return (float) $cart->items->sum(fn ($item) => $item->quantity * $item->unit_price);
    }

    private function authorizeCartItem(User $user, CartItem $cartItem): void
    {
        if ($cartItem->cart->user_id !== $user->id) {
            abort(403, 'Item keranjang ini bukan milik Anda.');
        }
    }

    private function availableStock(Product $product): int
    {
        $inventory = $product->inventory;

        return $inventory ? $inventory->available() : $product->stock;
    }

    private function unitPrice(Product $product, ?int $variantId): float
    {
        if ($variantId) {
            $variant = $product->variants()->findOrFail($variantId);

            return (float) $product->price + (float) $variant->price_adjustment;
        }

        return (float) $product->price;
    }
}

File: mobile/lib/services/cart_service.dart
import '../core/network/api_client.dart';
import '../models/cart_model.dart';

/// Service keranjang belanja — mengonsumsi /api/v1/cart.
class CartService {
  final _api = ApiClient();

  /// Ambil keranjang: GET /cart
  Future<CartModel> fetchCart() async {
    try {
      final res = await _api.get('/cart');
      return CartModel.fromJson(res.data['data'] ?? {'id': 0, 'items': []});
    } catch (_) {
      return CartModel(id: 0, items: []);
    }
  }

  /// Tambah item: POST /cart/items
  Future<void> addItem({required int productId, int quantity = 1, int? variantId}) async {
    await _api.post('/cart/items', data: {
      'product_id': productId,
      'quantity': quantity,
      if (variantId != null) 'variant_id': variantId,
    });
  }

  /// Update jumlah: PUT /cart/items/{id}
  Future<void> updateItem(int cartItemId, int quantity) async {
    await _api.put('/cart/items/$cartItemId', data: {
      'quantity': quantity,
    });
  }

  /// Hapus item: DELETE /cart/items/{id}
  Future<void> removeItem(int cartItemId) async {
    await _api.delete('/cart/items/$cartItemId');
  }

  /// Kosongkan keranjang: DELETE /cart
  Future<void> clear() async {
    await _api.delete('/cart');
  }
}
